cpp/src/arrow/compute/row/grouper.cc

```diff
@@ -329,6 +330,8 @@ Status CheckAndCapLengthForConsume(int64_t batch_length, int64_t& consume_offset
   return Status::OK();
 }
 
+enum class GrouperMode { kPopulate, kConsume, kLookup };
```

Comment:
Instead of having multiple grouper modes to instruct the only `ConsumeImpl`, do we want to organize the code the way that each "mode" has its own implementation, and each implementation can be a composition of a series of the underlying common primitive functions.

Of course this is not a strong bias, I'm just feeling that it might make the code more "plain".

Reply:
Same answer as to the other comment: let my try this out :)

Reply:
Hmm, I think this is going to get clumsy because the primitive functions would need to allocate and provide a number of temporary buffers (such as `offsets_batch` and `key_bytes_batch`). We could of course introduce dedicated structures to hold these.